### ISSUE TITLE
odd dialyzer warning with verify always

### ISSUE THREAD

**Issue by tsloughter** [tomas-abrahamsson/gpb#149]
So I'm dialyzing https://github.com/hexpm/hex_core while updating to proto3 and am baffled by a set of warnings that only popup when the gpb opts include `{verify, always}`:

```
src/hex_pb_signed.erl
 143: Guard test F@_1::binary() | maybe_improper_list() == '$undef' can never succeed
 146: Guard test F@_2::binary() | maybe_improper_list() == '$undef' can never succeed
 185: Guard test F@_1::binary() | maybe_improper_list() == '$undef' can never succeed
 188: Guard test F@_2::binary() | maybe_improper_list() == '$undef' can never succeed
```

143 and 146 refer to:

```erlang
dfp_read_field_def_Signed(<<>>, 0, 0, F@_1, F@_2, _) ->
    S1 = #{},
    S2 = if F@_1 == '$undef' -> S1;
	    true -> S1#{payload => F@_1}
	 end,
    if F@_2 == '$undef' -> S2;
       true -> S2#{signature => F@_2}
    end;
```
So it is in the decoding.

I do a diff between a build with verify always and verify optionally and it is just:

```
diff --git a/src/hex_pb_signed.erl b/src/hex_pb_signed.erl
index f3627ce..17c6156 100644
--- a/src/hex_pb_signed.erl
+++ b/src/hex_pb_signed.erl
@@ -41,10 +41,7 @@ encode_msg(Msg, MsgName) ->
 
 -spec encode_msg('Signed'(),'Signed', list()) -> binary().
 encode_msg(Msg, MsgName, Opts) ->
-    case proplists:get_bool(verify, Opts) of
-      true -> verify_msg(Msg, MsgName, Opts);
-      false -> ok
-    end,
+    verify_msg(Msg, MsgName, Opts),
     TrUserData = proplists:get_value(user_data, Opts),
     case MsgName of
       'Signed' -> e_msg_Signed(Msg, TrUserData)
```

So it, as expected, is only touching the encode function but the dialyzer failure is in decode!

Anyway, it isn't a big deal, I can just switch verify to optional. But it is such a weird issue that I can't explain I figured I'd open a ticket anyway :)

**Comment by tomas-abrahamsson** [comment]
The `if F@_1 == '$undef'` is indeed redundant for proto3, I think. So I think dialyzer is right in its analysis. I'll see if I can improve on code generation. I too, do not understand why setting verify to optional, changes dialyzer's output.

**Comment by tomas-abrahamsson** [comment]
I pushed to a branch, `fix-proto3-decoding-redundant-check` just now as an attempt to fix this dialyzer issue.

**Comment by tsloughter** [comment]
Cool,  yup, that fixes it :) thanks

**Comment by tomas-abrahamsson** [comment]
Super! I'll push it soon

**Comment by tomas-abrahamsson** [comment]
Included in 4.2.3 which I just pushed, so I'll be closing this one.
